# Lab Session 8 - Coursework Assignment #2
## Preliminaries
In this coursework you will analyse stocks comprising the Standard and Poor's 100 index. The assignment has been structured into separate exercises given below.

You will need to download a .csv-file containing daily adjusted closing prices for the components of the S&P 100 index. This data can be downloaded from the KEATS page.

You can then read this file into a Pandas DataFrame using code similar to that below.



## Import necessary libaries
- import numpy for scientific computing 
- import pandas for transforming csv to dataframe

In [4]:
import numpy as np
import pandas as pd

> First, I will check the current working path.

In [1]:
import os
current_path = os.getcwd()
print(current_path)

/Users/andrewyang/Desktop/workspace/Scientific-Computing-For-Finance/lab8


> And then download the S&P index csv "sap100.csv" and move it under the working path we got on the previous step.
> We'll check if the file really exists in the folder before doing futher analystics.

In [5]:
csv_path = current_path + '/sap100.csv'     # set the absolute path of the csv file
print(os.path.isfile(csv_path))             # Return true if the file exists

True


> 1. read the csv file we downloaded and checked last step by the function `pd.read_csv`
> 2. save the dataframe as `prices`
> 3. take a look of the dataframe `prices` to check if the indexs and columns are right by using the function `head()`

In [6]:
prices = pd.read_csv(csv_path)              # read the s&p 100 csv file and store it in a dataframe
prices.head()                               # check the dataframe

Date       AAPL         VZ        WBA        WFC        WMT  \
0  2016-06-06 00:00:00  24.657499  50.709999  78.660004  50.490002  71.050003   
1  2016-06-07 00:00:00  24.757500  51.750000  79.120003  50.270000  71.029999   
2  2016-06-08 00:00:00  24.735001  51.520000  79.440002  50.000000  71.279999   
3  2016-06-09 00:00:00  24.912500  51.950001  79.010002  49.139999  71.089996   
4  2016-06-10 00:00:00  24.707500  52.669998  82.470001  48.340000  71.139999   

         XOM        ABT       ADBE        AIG  ...       TMUS       TSLA  \
0  89.339996  39.169998  99.120003  57.290001  ...  43.070000  14.712000   
1  90.709999  38.930000  98.519997  57.060001  ...  43.740002  15.489333   
2  90.790001  38.910000  98.680000  56.919998  ...  43.910000  15.701333   
3  90.669998  38.869999  98.070000  56.099998  ...  43.080002  15.290667   
4  89.980003  38.400002  97.089996  54.990002  ...  41.930000  14.586000   

         TXN         UNH        UNP         UPS        USB          V  \
0  61.290001  138.149994  86.849998  103.110001  42.880001  80.540001   
1  61.560001  136.940002  87.570000  103.870003  42.660000  80.599998   
2  62.080002  140.399994  89.040001  104.690002  42.810001  81.650002   
3  62.139999  140.679993  89.389999  104.970001  42.540001  81.260002   
4  61.910000  139.240005  88.779999  104.519997  41.959999  80.180000   

        ABBV         ACN  
0  62.820000  119.599998  
1  63.099998  119.650002  
2  61.730000  119.320000  
3  60.750000  119.440002  
4  61.000000  118.269997  

[5 rows x 102 columns]

> 1. We can see the dataframe has 102 columns, which is the fist 2 column `index` and `Date` with 100 stocks.
> 2. The Date type is "YYYY-MM-DD HH:MM:SS"
> 3. The price of the stocks are float numbers

## Exercise 1
Choose a single stock from the index, and compute the corresponding daily, monthly and annual log-returns. For each sampling frequency, test whether the returns are normally distributed. Include appropriate graphs in your final report and ensure that you clearly explain your analysis in your own words.



> My thoughts:
> 1. Choose a stock, which I decided to choose here is `Nvidia Corp`. Because I know more about this company and maybe having some basic knowledge of the company benifits the analyse.
> 2. The symbol for Nvidia Corp is `NVDA`
> 3. Select the column `Date` with `NVDA` and save it as `nvda_price`
> 

In [24]:
nvda_price = prices[['Date', 'NVDA']]


## Exercise 2
Compute the Jarque-Bera test statistic and 
p
-values for all stocks in the index over daily, monthly and annual frequencies. Summarise your results in a single graph which shows three side-by-side boxplots of the test statistic comparing: daily, monthly and annual return distributions. Comment on whether your results support the claim that stock returns exhibit aggregational Gaussianity (see, e.g., Cont 2001; reference below).

## Exercise 3
Download data for the S&P 100 index (symbol ^OEX), and repeat Exercise 1 with this data.

## Exercise 4
Estimate the 
α
 and 
β
 of each stock according to the single-index model using ordinary least squares regression over monthly excess returns, storing your results in a DataFrame. Produce a single graph showing side-by-side boxplots summarising the distribution of the alpha and beta coefficients over all stocks in the index.



## Exercise 5
Compute the covariance matrix of the components of the index using the monthly excess-returns. According to the single-index model (see Elton and Gruber 1997; reference below), the covariance of a given pair of stocks should be directly proportional to the product of the corresponding 
β
 values. Test this hypothesis, and summarise your results in a single graph. Provide a written explanation of your results, and a short critical discussion.